### ISSUE TITLE
API for casting PyProxy  / JsProxy to native types

### ISSUE THREAD

**Issue by rth** [pyodide/pyodide#1192]
Following  https://github.com/iodide-project/pyodide/pull/1167#issuecomment-772912630 on improving `PyProxy.deepCopyToJavascript` and  `PyProxy.shallowCopyToJavascript ` name, I think it might be better to continue that discussion in a separate issue rather than a PR.

**Comment by rth** [comment]
>  If you have a preference I can open a PR. Maybe even copyToJS, copy2Js, deepCopy2Js. Is it better to keep the word deep in there? What about shallowCopyToJs and copyToJs? I guess deepCopyToJs is my favorite?

Thinking more about this I'm not sure "copy" is the right term. Copy is 
> make a similar or identical version of; reproduce

while we are not copying, we are more converting or casting the from language to another. There is also the question of,
 - if you have a nested object and you can't convert one of these objects, you would likely still return a PyProxy (or fail I guess, I haven't checked what is done, or it could be an option). So the conversion to JS is not always possible, while the word "copy"  somewhat implies that it would never fail.
 - another question is whether we would eventually need a similar casting for JsProxy, in which case it would be good to think about the API for it (Maybe there is one already I haven't checked). There maybe something like `to_py(deep=False)` would make sense.

In terms of naming maybe we could  also take some inspiration from [PyO3](https://github.com/PyO3/pyo3) which has an API for converting between Python and Rust. For instance, the [into_py](https://docs.rs/pyo3/0.13.1/pyo3/conversion/trait.IntoPy.html) trait so maybe `JsProxy.into_py(deep=False)` would indeed be not so bad. And then I imagine similarly `PyProxy.intoJs` though I'm not sure practical it would be to pass `deep` as an argument?

**Comment by phorward** [comment]
> In terms of naming maybe we could also take some inspiration from [PyO3](https://github.com/PyO3/pyo3) which has an API for converting between Python and Rust. For instance, the [into_py](https://docs.rs/pyo3/0.13.1/pyo3/conversion/trait.IntoPy.html) trait so maybe `JsProxy.into_py(deep=False)` would indeed be not so bad. And then I imagine similarly `PyProxy.intoJs` though I'm not sure practical it would be to pass `deep` as an argument?

Hi Roman, this sounds good, and you're right, the function names are not well decided for now. As a Rust-newbie, I'd also written several into_-functions and this is a good paradigm to borrow from. 

What about 

- `PyProxy.intoJs()` for deepCopyToJavascript and
- `PyProxy.asJs()` for shallowCopyToJavascript, as it is more like a cast (in Rust-terms `x as u32`)

The deep parameter would also be an option for me.

**Comment by hoodmane** [comment]
> we are not copying, we are more converting or casting the from language to another

An excellent point. The implicit conversion is like a cast, these "copy" APIs are converting.

> if you have a nested object and you can't convert one of these objects, you would likely still return a PyProxy

Yes. Currently `deepCopyToJavascript` does exactly what the old conversion function did implicitly. I have not considered at all whether or not this is the right behavior, we could consider throwing an error or something. 

Also, it's worth mentioning that the current treatment of tuples is a bit weird. It's logical from inside the implementation but may be surprising to users. We might want to treat tuples more like lists.

> In terms of naming maybe we could also take some inspiration from PyO3 which has an API for converting between Python and Rust. For instance, the into_py trait so maybe JsProxy.into_py(deep=False) would indeed be not so bad. 

Sounds great to me. In general I think Rust has very good naming conventions.

> And then I imagine similarly `PyProxy.intoJs` though I'm not sure practical it would be to pass deep as an argument?

Should be fine, we can give it a default value.

> `PyProxy.asJs()` for `shallowCopyToJavascript`

I think this isn't the right name because `asJs` is the implicit behavior, whereas `shallowCopy` does one layer of conversion.

**Comment by hoodmane** [comment]
Also, it's worth mentioning that the current treatment of tuples is a bit weird. It's logical from inside the implementation but may be surprising to users. We might want to treat tuples more like lists.

**Comment by hoodmane** [comment]
> another question is whether we would eventually need a similar casting for JsProxy, in which case it would be good to think about the API for it (Maybe there is one already I haven't checked).

We don't have such a thing, I am planning to make one when I get around to it. As much as possible I think we should try to make the API for conversions from Python to Javascript symmetric with the API for conversion from Javascript to Python.

**Comment by rth** [comment]
Sounds good. Let's then go with `PyProxy.intoJs` with a deep param if there is a consensus on that.

I'm also not too sure about `asJS` as the difference with `intoJS` would not be very explicit.

**Comment by hoodmane** [comment]
One question though: currently I accept a `depth` parameter. When `deep` is `true` that's equivalent to infinite `depth`, when `deep` is `false` that's equivalent to a `depth` of 1. `depth` equal to zero is a noop. Perhaps it's better to keep this more complicated API under a different name?

A second question: because the conversion function sometimes returns a proxy and other times returns a primitive type, this is likely to lead to code like:

```javascript
let result = runPython(code);
if(result.intoJs){
   let jsresult = result.intoJs();
   result.destroy();
   result = jsresult;
}
```
Perhaps a static version `PyProxy.intoJs` which returns without altering objects that aren't a `PyProxy` would help, so then we could say more like:
```javascript
let result = PyProxy.intoJs(runPython(code));
```
Of course then there is the question of when/whether to destroy the input proxy...

**Comment by hoodmane** [comment]
Also: because the conversion function sometimes returns a proxy and other times returns a primitive type, this is likely to lead to code like:

```javascript
let result = runPython(code);
if(result.intoJs){
   let jsresult = result.intoJs();
   result.destroy();
   result = jsresult;
}
```
Perhaps a static version `PyProxy.intoJs` which returns without altering objects that aren't a `PyProxy` would help, so then we could say more like:
```javascript
let result = PyProxy.intoJs(runPython(code));
```
Of course then there is the question of when/whether to destroy the input proxy...

**Comment by rth** [comment]
> Currently it accepts a depth parameter. When deep is true that's equivalent to infinite depth, when deep is false that's equivalent to a depth of 1. depth equal to zero is a noop. Perhaps it's better to keep this more complicated API under a different name?

Maybe then only have `depth` and suggest to use `depth=1` for shallow copies? It's a bit more technical, but not that much. Not sure.

> let result = PyProxy.intoJs(runPython(code));

It could work. Tough I would have expected more something like `<some_class>.from(runPython(code)).intoJs()`. It's strange to both do a constructor and a converstion with the static version of `intoJs`. Though it might be harder to implement. Aren't there potentially other `PyProxy` methods we might want to use on the result of `runPython` in the future?

**Comment by daoxian** [comment]
> * another question is whether we would eventually need a similar casting for JsProxy, in which case it would be good to think about the API for it (Maybe there is one already I haven't checked). There maybe something like `to_py(deep=False)` would make sense.

The API  JsProxy.to_py(deep=False) would cause a really big performance improvement, I think, since from time to time in python we have to aquire arrays returned from JS methods,  i.e.,  a huge image array returned by some method of opencv.js, which is now lacked in python packages.

**Comment by hoodmane** [comment]
@daoxian I'll open a separate issue to talk about buffer problems, the buffers need special handling.

**Comment by hoodmane** [comment]
Okay I created two separate issues #1202 and #1203. The underlying issues in the two directions are different and require different fixes so I figured it made more sense to keep them separate.

**Comment by hoodmane** [comment]
I believe this is resolved by #1231 and #1244.
